### PR TITLE
NIFI-14077 Add ProcessGroup Performance Metrics to Prometheus

### PR DESCRIPTION
**Pull request by esecules**
# Summary

[NIFI-14077](https://issues.apache.org/jira/browse/NIFI-14077)
- Add ProcessGroup Performance Metrics to Prometheus if performance metric collection is available
- Add total task duration metric to each process group, not just the root process group
- Add allowable values to the API docs for the producer and includedRegistries parameters so users know what there is to query without going to the source code.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing
(no new deps)
- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Comment by esecules** [comment]
I suspect the macos system test is flakey so I am rerunning it on my fork https://github.com/esecules/nifi/actions/runs/12285436758?pr=3

**Comment by exceptionfactory** [comment]
> I suspect the macos system test is flakey so I am rerunning it on my fork https://github.com/esecules/nifi/actions/runs/12285436758?pr=3
> 
> EDIT: The run passed on my fork hopefully it does the same below (thanks for the speedy retry @exceptionfactory!)

You're welcome, I have recently noticed a couple flaky tests in the system-tests workflow.

**Comment by esecules** [comment]
Looks like the retry failed still 😔, however it's passing in my fork and my machine (MacBook).

Do these tests run in parallel and might interfere with each other?

Or the tests might be running on an underpowered server and the 60 second timeout isn't sufficient for GitHub Actions?

**Comment by esecules** [comment]
Looks like system tests passed on retry! Thanks for all the retries.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for proposing these changes @esecules. I'm not complete sure about introducing the additional gauges because the processing performance information can be disabled. However, it looks like the null handling accounts for that fact, so this approach may be sufficient. I noted a few minor recommendations.

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java [hunk 1]
```diff
@@ -275,5 +275,35 @@ public NiFiMetricsRegistry() {
                 .help("Provenance repository free space in bytes")
                 .labelNames("instance", "component_type", "component_name", "component_id", "parent_id", "repo_identifier")
                 .register(registry));
+
+        nameToGaugeMap.put("PROCESSING_PERF_CPU_MILLIS", Gauge.build()
+                .name("nifi_processing_cpu_duration")
+                .help("Estimated cpu time (in milliseconds) used by this component")
```
Comment (exceptionfactory) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java:281]:
```suggestion
                .help("Estimated CPU time (in milliseconds) used by this component")
```

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java [hunk 2]
```diff
@@ -275,5 +275,35 @@ public NiFiMetricsRegistry() {
...
+                .labelNames("instance", "component_type", "component_name", "component_id", "parent_id", "repo_identifier")
+                .register(registry));
+
+        nameToGaugeMap.put("PROCESSING_PERF_GC_MILLIS", Gauge.build()
+                .name("nifi_processing_gc_duration")
+                .help("Estimated gc time (in milliseconds) used by this component")
```
Comment (exceptionfactory) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java:287]:
```suggestion
                .help("Estimated garbage collection time (in milliseconds) used by this component")
```

**Comment by esecules** [comment]
> Thanks for proposing these changes @esecules. I'm not complete sure about introducing the additional gauges because the processing performance information can be disabled. However, it looks like the null handling accounts for that fact, so this approach may be sufficient. I noted a few minor recommendations.

I'll check if I already added test coverage for when the performance feature is disabled and I'll add it if it's missing.

**Comment by esecules** [comment]
> > Thanks for proposing these changes @esecules. I'm not complete sure about introducing the additional gauges because the processing performance information can be disabled. However, it looks like the null handling accounts for that fact, so this approach may be sufficient. I noted a few minor recommendations.
> 
> I'll check if I already added test coverage for when the performance feature is disabled and I'll add it if it's missing.


I've looked over the testing for this module. I don't see any interactions with settings so I'll add a test case to make sure we continue to handle null performance information.

**Comment by esecules** [comment]
Added the tests which validate that we don't blow up when the performance status is null.

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/FlowResource.java [hunk 3]
```diff
@@ -567,12 +567,14 @@ public Response getBreadcrumbs(
     )
     public Response getFlowMetrics(
             @Parameter(
-                    description = "The producer for flow file metrics. Each producer may have its own output format.",
+                    description = "The producer for flow file metrics. Each producer may have its own output format. " +
+                            "Allowed values: [prometheus, json]",
```
[the remark this answers is not in the document]
Comment (esecules) [reply nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/FlowResource.java:558]: my bad i forgot to commit this change. brb.

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/FlowResource.java [hunk 4]
```diff
@@ -555,11 +555,15 @@ public Response getBreadcrumbs(
...
+                    required = true,
+                    schema = @Schema(allowableValues = {"prometheus", "json"})
             )
             @PathParam("producer") final String producer,
             @Parameter(
-                    description = "Set of included metrics registries"
+                    description = "Set of included metrics registries. Duplicate the parameter to include multiple registries. " +
+                            "All registries are included by default.",
+
+                    schema = @Schema(allowableValues = {"NIFI", "JVM", "BULLETIN", "CONNECTION", "CLUSTER"})
```
Comment (esecules) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/FlowResource.java:566]: Without the schema / allowableValues, the generated swagger.yaml has all these values already, but either way doesnt make a difference to the final API in my browser. It seems to get them from inspecting the type of this parameter. The same can't be said for the `producer` above since its just a string. However, I am not seeing any of these parameters in the `/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/target/nifi-web-api-2.2.0-SNAPSHOT/docs/rest-api/index.html` file when i load it in my browser. Maybe whatever generates that from the swagger spec is missing something?

There's also an existing API in the same scenario, it's swagger has the allowed enum values listed, but they do not show up in the docs:
image
image

Comment (exceptionfactory) [reply nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/FlowResource.java:566]: That HTML rendering is from version 1, NiFi 2 uses a simpler HTML template that should include the allowable values. As long as the enumerated values show up in the YAML, we can address the HTML rendering separately.

Comment (esecules) [reply nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/FlowResource.java:566]: Good to know. The YAML looks good from my side. Looking to see the v2 API docs. I'll resolve this comment.

**Comment by esecules** [comment]
Hi @exceptionfactory I believe I resolved your comments. Could you let me know if there's anything else?

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java [hunk 5]
```diff
@@ -275,5 +275,35 @@ public NiFiMetricsRegistry() {
                 .help("Provenance repository free space in bytes")
                 .labelNames("instance", "component_type", "component_name", "component_id", "parent_id", "repo_identifier")
                 .register(registry));
+
+        nameToGaugeMap.put("PROCESSING_PERF_CPU_MILLIS", Gauge.build()
```
Comment (exceptionfactory) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java:279]:
```suggestion
        nameToGaugeMap.put("PROCESSING_PERFORMANCE_CPU_DURATION", Gauge.build()
```

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java [hunk 6]
```diff
@@ -275,5 +275,35 @@ public NiFiMetricsRegistry() {
...
+                .name("nifi_processing_performance_cpu_duration")
+                .help("Estimated CPU time (in milliseconds) used by this component")
+                .labelNames("instance", "component_type", "component_name", "component_id", "parent_id", "repo_identifier")
+                .register(registry));
+
+        nameToGaugeMap.put("PROCESSING_PERF_GC_MILLIS", Gauge.build()
```
Comment (exceptionfactory) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java:285]:
```suggestion
        nameToGaugeMap.put("PROCESSING_PERFORMANCE_GC_DURATION", Gauge.build()
```

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java [hunk 7]
```diff
@@ -275,5 +275,35 @@ public NiFiMetricsRegistry() {
...
+                .name("nifi_processing_performance_gc_duration")
+                .help("Estimated garbage collection time (in milliseconds) used by this component")
+                .labelNames("instance", "component_type", "component_name", "component_id", "parent_id", "repo_identifier")
+                .register(registry));
+
+        nameToGaugeMap.put("PROCESSING_PERF_READ_MILLIS", Gauge.build()
```
Comment (exceptionfactory) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java:291]:
```suggestion
        nameToGaugeMap.put("PROCESSING_PERFORMANCE_CONTENT_READ_DURATION", Gauge.build()
```

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java [hunk 8]
```diff
@@ -275,5 +275,35 @@ public NiFiMetricsRegistry() {
...
+                .name("nifi_processing_performance_content_read_duration")
+                .help("Estimated content read time (in milliseconds) used by this component")
+                .labelNames("instance", "component_type", "component_name", "component_id", "parent_id", "repo_identifier")
+                .register(registry));
+
+        nameToGaugeMap.put("PROCESSING_PERF_WRITE_MILLIS", Gauge.build()
```
Comment (exceptionfactory) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java:297]:
```suggestion
        nameToGaugeMap.put("PROCESSING_PERFORMANCE_CONTENT_WRITE_DURATION", Gauge.build()
```

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java [hunk 9]
```diff
@@ -275,5 +275,35 @@ public NiFiMetricsRegistry() {
...
+                .name("nifi_processing_performance_content_write_duration")
+                .help("Estimated content write time (in milliseconds) used by this component")
+                .labelNames("instance", "component_type", "component_name", "component_id", "parent_id", "repo_identifier")
+                .register(registry));
+
+        nameToGaugeMap.put("PROCESSING_PERF_COMMIT_MILLIS", Gauge.build()
```
Comment (exceptionfactory) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/NiFiMetricsRegistry.java:303]:
```suggestion
        nameToGaugeMap.put("PROCESSING_PERFORMANCE_SESSION_COMMIT_DURATION", Gauge.build()
```

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/PrometheusMetricsUtil.java [hunk 10]
```diff
@@ -508,4 +518,25 @@ private static void addStorageUsageMetric(final NiFiMetricsRegistry nifiMetricsR
...
+                    instanceId, componentType, componentName, componentId, parentId, perfStatus.getIdentifier());
+
+            niFiMetricsRegistry.setDataPoint(perfStatus.getContentReadDuration() / 1000.0, "PROCESSING_PERF_READ_MILLIS",
+                    instanceId, componentType, componentName, componentId, parentId, perfStatus.getIdentifier());
+
+            niFiMetricsRegistry.setDataPoint(perfStatus.getContentWriteDuration() / 1000.0, "PROCESSING_PERF_WRITE_MILLIS",
+                    instanceId, componentType, componentName, componentId, parentId, perfStatus.getIdentifier());
+
+            niFiMetricsRegistry.setDataPoint(perfStatus.getSessionCommitDuration() / 1000.0, "PROCESSING_PERF_COMMIT_MILLIS",
+                    instanceId, componentType, componentName, componentId, parentId, perfStatus.getIdentifier());
```
Comment (exceptionfactory) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/PrometheusMetricsUtil.java:539]: The `PROCESSING_PERF` references should be updated to align with the recommendations in the Metrics Registry.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for your patience on this @esecules. I noted one more set of recommendations to align the registry metric name with the gauge name. Otherwise this looks close to completion.

**Comment by esecules** [comment]
> Thanks for your patience on this @esecules. I noted one more set of recommendations to align the registry metric name with the gauge name. Otherwise this looks close to completion.

I'll get on it. As always thanks for the review.

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/PrometheusMetricsUtil.java (hunk 10 above)
Comment (esecules) [reply nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/prometheusutil/PrometheusMetricsUtil.java]: done

**Review by exceptionfactory — APPROVED** [review]
Thanks for working through the feedback @esecules, the latest version looks good! +1 merging
